File: 2020/21/main_test.go
package main

import (
	"os"
	"testing"
)

func readTestInput() []Food {
	file, _ := os.Open("input_test.txt")
	defer file.Close()
	return readInput(file)
}

func TestShouldReadInput(t *testing.T) {
	input := readTestInput()
	expected := Food{
		Ingredients: []string{"trh", "fvjkl", "sbzzf", "mxmxvkd"},
		Allergens:   []string{"dairy"},
	}

	var found bool
loop:
	for _, in := range input {
		if len(expected.Ingredients) != len(in.Ingredients) {
			continue
		}
		if len(expected.Allergens) != len(in.Allergens) {
			continue
		}
		found = true
		for i := range expected.Ingredients {
			if in.Ingredients[i] != expected.Ingredients[i] {
				found = false
				continue loop
			}
		}
	}
	if !found {
		t.Errorf("did not read input correctly, did not find expected")
	}
}

func TestTask1ShouldFindResult(t *testing.T) {
	input := readTestInput()
	expect := 5

	result := Task1(input, true)
	if result != expect {
		t.Errorf("got %d, expected %d", result, expect)
	}
}

func TestTask2ShouldFindResult(t *testing.T) {
	input := readTestInput()
	expect := "mxmxvkd,sqjhc,fvjkl"

	result := Task2(input, true)
	if result != expect {
		t.Errorf("got %s, expected %s", result, expect)
	}
}

File: 2021/07/src/main_test.zig
const std = @import("std");

const app = @import("./main.zig");

test "expect task 1 to result in 37 fuel" {
    var input = [_]u32{
        16, 1, 2, 0, 4, 2, 7, 1, 2, 14,
    };
    const expected: u32 = 37;

    try std.testing.expect(app.task1(&input) == expected);
}

test "expect task 2 to result in 168 fuel" {
    var input = [_]u32{
        16, 1, 2, 0, 4, 2, 7, 1, 2, 14,
    };
    const expected: u32 = 168;

    try std.testing.expect(app.task2(&input) == expected);
}
